### PR TITLE
Start making each `FormatterStep` roundtrip serializable.

### PR DESCRIPTION
**Pull request by nedtwigg**
I don't like the design of the Gradle configuration cache. It puts too much burden on plugin developers, and it's too slow for end users (#987). But our workaround is a bit of a hassle, and we've had a really active contributor pool lately so maybe we have the brainpower to do the big rewrite that it would take to remove the `JvmLocalCache` workaround and fully support roundtrip serialization.

This PR introduces `StepHarnessBase`, which is a mechanism for enforcing that every individual formatter step must support roundtrip serialization. It will take a while for every step to support it, so we can adjust this `StepHarnessBase` as more and more steps add support. Once every step supports it, we will enforce this requirement for all new steps.

This PR also introduces `FormatterStepEqualityOnStateSerialization`. This allows us to keep the exact same equality semantics we have used so far, but it gives us more flexibility to implement serialization. Especially serialization of absolute paths without screwing up buildcache keys.

### CODE REVIEW

**Comment by nedtwigg** [comment]
@Goooler does this approach make sense? A nice thing is that we could merge these bit by bit, we don't have a hard fork situation, and once we have completed [all of these](https://github.com/diffplug/spotless/issues/1274#issuecomment-1839694669) we can remove `JvmLocalCache` and we'll have full native support for Gradle's configuration cache.

**Comment by jbduncan** [comment]
This PR looks fantastic! Let me know when you're close to the end and I'd be more than happy to give this a proper review.

**Review by Goooler — APPROVED** [review]

**Comment by jbduncan** [comment]
@nedtwigg It took me all evening to get my machine set up for Spotless again (fixing my vastly outdated fork on GitHub which was still using `master`, for instance), so I've not had a chance to review the code yet. Maybe next weekend!

**Comment by Goooler** [comment]
Use `gh pr checkout 1945`.

**Comment by jbduncan** [comment]
Thanks for the pointer, @Goooler! I've already fixed my fork and run `git checkout serializable-refactor`, but I'll consider `gh` for the future. 👍

File: lib/src/main/java/com/diffplug/spotless/FileSignature.java [hunk 1]
```diff
@@ -93,6 +96,35 @@ private FileSignature(final List<File> files) throws IOException {
 		}
 	}
 
+	/** A view of `FileSignature` which can be safely roundtripped. */
+	public static class Promised implements Serializable {
+		private static final long serialVersionUID = 1L;
+		private final List<File> files;
```
[the remark this answers is not in the document]
Comment (nedtwigg) [reply lib/src/main/java/com/diffplug/spotless/FileSignature.java:102]: I don't have much preference for `String` vs `File`, I guess `File` is just a tiny hint that it's not a phone number or name or something, it's a File. Do you see some downsides to `File` vs `String`?

File: lib/src/main/java/com/diffplug/spotless/FormatterStep.java [hunk 2]
```diff
@@ -104,6 +104,51 @@ public final String format(String rawUnix, File file) throws Exception {
...
+	 *             If the step has any state, this supplier will calculate it lazily. The supplier doesn't
+	 *             have to be serializable, but the result it calculates needs to be serializable.
+	 * @param equalityFunc
+	 * 		       A pure serializable function (method reference recommended) which takes the result of `roundtripInit`,
+	 * 		       and returns a serializable object whose serialized representation will be used for `.equals` and
+	 * 		       `.hashCode` of the FormatterStep.
+	 * @param formatterFunc
+	 * 		       A pure serializable function (method reference recommended) which takes the result of `equalityFunc`,
+	 * 		       and returns a `FormatterFunc` which will be used for the actual formatting.
+	 * @return A FormatterStep which can be losslessly roundtripped through the java serialization machinery.
```
[the remark this answers is not in the document]
Comment (nedtwigg) [reply lib/src/main/java/com/diffplug/spotless/FormatterStep.java:120]: I don't see what's broken abut it. Whitespace? Or something about the content? There are no warnings in my IntelliJ.

File: lib/src/main/java/com/diffplug/spotless/FormatterStep.java [hunk 3]
```diff
@@ -104,6 +104,51 @@ public final String format(String rawUnix, File file) throws Exception {
...
+	 * @param roundTrip
+	 *             The roundtrip serializable state of the step.
+	 * @param equalityFunc
+	 * 		       A pure serializable function (method reference recommended) which takes the result of `roundTrip`,
+	 * 		       and returns a serializable object whose serialized representation will be used for `.equals` and
+	 * 		       `.hashCode` of the FormatterStep.
+	 * @param formatterFunc
+	 * 		       A pure serializable function (method reference recommended) which takes the result of `equalityFunc`,
+	 * 		       and returns a `FormatterFunc` which will be used for the actual formatting.
+	 * @return A FormatterStep which can be losslessly roundtripped through the java serialization machinery.
```
[the remark this answers is not in the document]
Comment (nedtwigg) [reply lib/src/main/java/com/diffplug/spotless/FormatterStep.java:142]: https://github.com/diffplug/spotless/pull/1945#discussion_r1428590002

File: lib/src/main/java/com/diffplug/spotless/FileSignature.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Goooler) [reply lib/src/main/java/com/diffplug/spotless/FileSignature.java:102]: I remembered there might be some serialization issues about using `File` in beans, not very sure. We can just keep `File` here.

File: lib/src/main/java/com/diffplug/spotless/FormatterStep.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Goooler) [reply lib/src/main/java/com/diffplug/spotless/FormatterStep.java:121]: `@param` and the description are not in the same line, intended to leave that?

File: lib/src/main/java/com/diffplug/spotless/JarState.java [hunk 4]
```diff
@@ -107,10 +147,4 @@ public ClassLoader getClassLoader() {
 	public ClassLoader getClassLoader(Serializable key) {
 		return SpotlessCache.instance().classloader(key, this);
 	}
-
-	/** Returns unmodifiable view on sorted Maven coordinates */
-	@Deprecated
-	public Set<String> getMavenCoordinates() {
-		return Collections.unmodifiableSet(mavenCoordinates);
-	}
```
Comment (jbduncan) [inline lib/src/main/java/com/diffplug/spotless/JarState.java:115]: I suppose it was already the plan to bump the major version of `lib` and mention the removal of this method in the changelog, wasn't it? :)

File: lib/src/main/java/com/diffplug/spotless/generic/IndentStep.java [hunk 5]
```diff
@@ -49,32 +55,21 @@ public FormatterStep create(int numSpacesPerTab) {
 
 	/** Creates a step which will indent with the given type of whitespace, converting between tabs and spaces at the given ratio. */
 	public static FormatterStep create(Type type, int numSpacesPerTab) {
-		Objects.requireNonNull(type, "type");
 		return FormatterStep.create("indentWith" + type.tabSpace("Tabs", "Spaces"),
```
Comment (jbduncan) [inline lib/src/main/java/com/diffplug/spotless/generic/IndentStep.java:58]: My IDE tells me `type.tabSpace("Tabs", "Spaces")` can be simplified to `type.tabSpace()` if `tabSpace()` is simplified to:

```java
private String tabSpace() {
	return this == TAB ? "Tabs" : "Spaces";
}
```

WDYT?

File: lib/src/main/java/com/diffplug/spotless/generic/IndentStep.java [hunk 6]
```diff
@@ -49,32 +55,21 @@ public FormatterStep create(int numSpacesPerTab) {
 
 	/** Creates a step which will indent with the given type of whitespace, converting between tabs and spaces at the given ratio. */
 	public static FormatterStep create(Type type, int numSpacesPerTab) {
-		Objects.requireNonNull(type, "type");
```
Comment (jbduncan) [inline lib/src/main/java/com/diffplug/spotless/generic/IndentStep.java:52]: At first I wasn't sure about this, but then I remembered Java 14+ has [JEP 358: Helpful NullPointerExceptions](https://openjdk.org/jeps/358), and I'm not sure if punishing Java 14+ users with less helpful error messages like what we had here is worth the small drop in error message quality for users stuck on older versions.

**tl;dr**: This seems fine to me!

**Review by jbduncan — CHANGES_REQUESTED** [review]
Looking good, and I like the incremental approach to the new roundtrip serialization mechanism. A few things from me (see the review comments), but otherwise a fantastic start!

File: lib/src/main/java/com/diffplug/spotless/Formatter.java [hunk 7]
```diff
@@ -304,6 +304,8 @@ public void close() {
 		for (FormatterStep step : steps) {
 			if (step instanceof FormatterStepImpl.Standard) {
 				((FormatterStepImpl.Standard) step).cleanupFormatterFunc();
+			} else if (step instanceof FormatterStepEqualityOnStateSerialization) {
+				((FormatterStepEqualityOnStateSerialization) step).cleanupFormatterFunc();
```
[the remark this answers is not in the document]
Comment (Goooler) [reply lib/src/main/java/com/diffplug/spotless/Formatter.java:308]: It would be better to just let finished ones extends `AutoCloseable`, then convert them one by one.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -95,10 +95,31 @@ Here's a checklist for creating a new step for Spotless:
 
 - [ ] Class name ends in Step, `SomeNewStep`.
 - [ ] Class has a public static method named `create` that returns a `FormatterStep`.
-- [ ] Has a test class named `SomeNewStepTest`.
+- [ ] Has a test class named `SomeNewStepTest` that uses `StepHarness` or `StepHarnessWithFile` to test the step.
+  - [ ] Start with `StepHarness.forStep(myStep).supportsRoundTrip(false)`, and then add round trip support as decribed in the next section.
```
[the remark this answers is not in the document]
Comment (nedtwigg) [reply CONTRIBUTING.md:99]: Thanks!

File: lib/src/main/java/com/diffplug/spotless/generic/IndentStep.java (hunk 5 above)
Comment (nedtwigg) [reply lib/src/main/java/com/diffplug/spotless/generic/IndentStep.java:58]: What I like about the current implementation is that it's clear that the final values are

- indentWithTabs
- indentWithSpaces

If we performed this refactor, it might seem like it was `indentWithTab`. It's nice when values are declared in the same place they are used, especially if they're only used once.

File: lib/src/main/java/com/diffplug/spotless/FormatterStep.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nedtwigg) [reply lib/src/main/java/com/diffplug/spotless/FormatterStep.java:121]: Testing shows it's fine to be on a separate line. I think a bit easier to read when the parameters have whole paragraphs.

File: lib/src/main/java/com/diffplug/spotless/JarState.java (hunk 4 above)
Comment (nedtwigg) [reply lib/src/main/java/com/diffplug/spotless/JarState.java:115]: I appreciate the callout!

File: lib-extra/src/main/java/com/diffplug/spotless/extra/EclipseBasedStepBuilder.java [hunk 9]
```diff
@@ -187,12 +186,6 @@ public Properties getPreferences() {
 			return preferences.getProperties();
 		}
 
-		/** Returns first coordinate from sorted set that starts with a given prefix.*/
-		public Optional<String> getMavenCoordinate(String prefix) {
-			return jarState.getMavenCoordinates().stream()
-					.filter(coordinate -> coordinate.startsWith(prefix)).findFirst();
-		}
-
```
[the remark this answers is not in the document]
Comment (nedtwigg) [reply lib-extra/src/main/java/com/diffplug/spotless/extra/EclipseBasedStepBuilder.java:195]: I very much appreciate the callout!

File: lib/src/main/java/com/diffplug/spotless/Formatter.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (nedtwigg) [reply lib/src/main/java/com/diffplug/spotless/Formatter.java:308]: Probably worthwhile, but mostly a separate concern from roundtrip serialization.

- tracking at #2017

**Comment by nedtwigg** [comment]
Thanks very much for this thorough review! My plan is

- deploy main as-is ([in progress](https://github.com/diffplug/spotless/actions/runs/7622153489/job/20759717126))
- merge this and its follow-ons
- definitely allow breaking changes in lib
- hoping no breaking changes in plugin-maven or plugin-gradle, but TBD

Until we have complete switched over to configuration cache, I'd rather not publish any releases. That creates a bit of an awkward phase, but I think we can keep it down to just one or two weeks. Can always backtrack and make some kind of bugfix release branch if we have to.

File: lib/src/main/java/com/diffplug/spotless/FileSignature.java [hunk 10]
```diff
@@ -93,6 +96,35 @@ private FileSignature(final List<File> files) throws IOException {
...
+				// null when restored via serialization
+				cached = ThrowingEx.get(() -> new FileSignature(files));
+			}
+			return cached;
+		}
+	}
+
+	public static Promised promise(Iterable<File> files) {
+		return new Promised(MoreIterables.toNullHostileList(files), null);
+	}
```
[the remark this answers is not in the document]
Comment (nedtwigg) [reply lib/src/main/java/com/diffplug/spotless/FileSignature.java:122]: Good call!

**Comment by jbduncan** [comment]
LGTM! 👍
